aten/src/ATen/cuda/CUDAContext.cpp

```diff
@@ -1,19 +1,48 @@
 #include "ATen/cuda/CUDAContext.h"
 #include "THC/THCGeneral.hpp"
 
+#include <mutex>
+#include <deque>
+#include <vector>
+
 namespace at { namespace cuda {
 
+namespace {
+
+static DeviceIndex num_gpus = -1;
```

Comment:
Hmm, aren't we caching this in CUDAStream.cpp as well? Well... I guess it's harmless, as long as we never want to support gpu hotplugging (lol)

Reply:
C++ nit: if you put it in an anonymous namespace, you don't have to also declare it `static`.

Reply:
Thanks for the review! Regarding the first point, the cached static variables shouldn't conflict right? since the variables are in anonymous namespace in both CUDAStream.cpp and CUDAContext.cpp? On a different note, THCState has `numDevices` which gets cached during THCudaInit. However, THCState_getNumDevices is not initialized anywhere in pytorch. So I was wondering if we should be caching numGPUs for calls like CUDAHooks::getNumGPUs() or c10::cuda::device_count().

I have removed static from entities in the anonymous namespace.